feat: Project transform

marvinlanhenke opened the pull request:
### Which issue does this PR close?
Closes #264 

### Rationale for this change
The ability to project row_filter to `Transform` partition. 
This will unblock the Manifest & PartitionEvaluator, which will enable the pruning of manifest files in `fn plan_files`.

### What changes are included in this PR?
- add `fn project(...)` on `Transform`
- mainly based on [Java ](https://github.com/apache/iceberg/tree/d350c9b8c995a2953aa8b80a0a1fc7cadc4dd16a/api/src/main/java/org/apache/iceberg/transforms) implementation

### Are these changes tested?
Yes. Unit tests are included.

marvinlanhenke commented:
@liurenjie1024 @ZENOTME @sdd PTAL

sdd commented on crates/iceberg/src/spec/transform.rs:
```diff
@@ -398,6 +679,80 @@ mod tests {
+
+        assert_eq!(format!("{}", result_unary), "projected_name IS NULL");
+        assert_eq!(format!("{}", result_binary), "projected_name = 0");
+        assert_eq!(format!("{}", result_set), "projected_name IN (0)");
```
Just trying to follow what's happening here so that I understand. 

So in the case of `result_binary`, the value of 5 gets truncated to 0, since the truncate transform when applied to an int effectively divides by 10, rounding down for all fractional values (ie, 1-9 get truncated to 0, 10-19 get truncated to 1, 20-29 get truncated to 2)?

And in the case of the set, since both 5 and 6 get truncated to the same value of 0, a set of (5, 6) becomes a set of (0)?

marvinlanhenke replied:
Yes, I think your understanding is correct.

Fokko replied:
I think it is better to have more than one element in the set. In Python and Java a `IN (0)` is being rewritten to `= 0` before evaluated.

Fokko replied:
Example can be found here: https://github.com/apache/iceberg/blob/81b62c78e0c230516090becda7d6040ee03e6a91/api/src/test/java/org/apache/iceberg/transforms/TestTruncatesProjection.java#L189-L190

It would be good to port the other tests of Java as well.

sdd reviewed (approved):
LGTM, with the caveat that I'm not an expert on this part of the spec, but the PR seems well-structured.

I wouldn't mind seeing more comments in the tests though to explain the "why" aspect of each test.

marvinlanhenke commented:
> LGTM, with the caveat that I'm not an expert on this part of the spec, but the PR seems well-structured.
> 
> I wouldn't mind seeing more comments in the tests though to explain the "why" aspect of each test.

Thanks for your feedback. If the others approve "the correctness" of the PR - I'll add those comments.

Fokko commented on crates/iceberg/src/spec/transform.rs:
```diff
@@ -398,6 +679,80 @@ mod tests {
+    }
+
+    #[test]
+    fn test_truncate_project() -> Result<()> {
```
Could you port all the tests from Java: https://github.com/apache/iceberg/blob/main/api/src/test/java/org/apache/iceberg/transforms/TestTruncatesProjection.java

Fokko commented on crates/iceberg/src/spec/transform.rs:
```diff
@@ -398,6 +679,80 @@ mod tests {
+
+        Ok(())
+    }
+
```
I think the most important tests are missing around the dates projection: https://github.com/apache/iceberg/blob/main/api/src/test/java/org/apache/iceberg/transforms/TestDatesProjection.java

Fokko commented on crates/iceberg/src/spec/transform.rs:
```diff
@@ -398,6 +679,80 @@ mod tests {
+    }
+
+    #[test]
+    fn test_bucket_project() -> Result<()> {
```
Can you port the test suite from Java to Rust: https://github.com/apache/iceberg/blob/main/api/src/test/java/org/apache/iceberg/transforms/TestBucketingProjection.java

marvinlanhenke replied:
@Fokko @liurenjie1024 
...I ported the suite for `Bucket` transforms - with all tests passing. So bucket impl seems to produce the correct results (at least no obvious bugs present). 

However, PTAL.

Fokko reviewed (commented):
Thanks for picking this up @marvinlanhenke 

The most important part of adding projection is making that they are absolutely correct. If rust would generate something different than Java of Python, leads to data incorrectness since it would not correctly evaluate the partition predicates.

marvinlanhenke commented:
> Thanks for picking this up @marvinlanhenke
> 
> The most important part of adding projection is making that they are absolutely correct. If rust would generate something different than Java of Python, leads to data incorrectness since it would not correctly evaluate the partition predicates.

@Fokko 
Thanks for the extensive review. Not only did I miss some test but the complete implementation for `Dates` : https://github.com/apache/iceberg/blob/d350c9b8c995a2953aa8b80a0a1fc7cadc4dd16a/api/src/main/java/org/apache/iceberg/transforms/Dates.java#L123 // I was just looking for the transforms for Years, Months, etc. So thank you for hinting me at this - I'll implement this as well - and try to port the Java Testsuite.

marvinlanhenke commented:
@Fokko 
Is it correct to assume, as far as I understood the Java implementation, that we support `dates` projection only for `year, month, and day`?

@liurenjie1024
I converted this to a draft - since I had to change the design. I'll push this draft, so you can verify. Basically, I got rid of the trait and implemented `boundary` on `Datum` itself - which makes more sense to me, now that I have a better overall picture with the dates transformations in mind.

liurenjie1024 reviewed (commented):
Thanks @marvinlanhenke for picking up this. In general it looks good, but I have some small suggestions:

1. How about we split this pr into smaller prs where each pr implements project for one transform?
2. How about we move specific logic of transform into each file in transform module?
3. I agree with @Fokko that we should port tests from java to ensure correctness.

liurenjie1024 commented:
> Basically, I got rid of the trait and implemented boundary on Datum itself - which makes more sense to me, now that I have a better overall picture with the dates transformations in mind.

@marvinlanhenke Thanks for this, I also an boundary trait is a little over design, and implementing them directly on `Datum` seems better to me.

marvinlanhenke commented on crates/iceberg/src/transform/mod.rs:
```diff
@@ -16,6 +16,7 @@
 // under the License.
 
 //! Transform function used to compute partition values.
+
```
you mean the comment? or the extra whitespace?

liurenjie1024 replied:
The extra whitespace.

liurenjie1024 commented on crates/iceberg/src/spec/transform.rs:
```diff
@@ -261,6 +270,174 @@ impl Transform {
+    /// value should match `bucket(10, 37) as bs`, and we project `a = 10` to
+    /// `bs = bucket(10, 37)`
+    pub fn project(&self, name: String, predicate: &BoundPredicate) -> Result<Option<Predicate>> {
+        let func = create_transform_function(self)?;
```
Yes, exactly.

marvinlanhenke commented on crates/iceberg/src/spec/transform.rs:
```diff
@@ -261,6 +270,174 @@ impl Transform {
+                func.transform_literal(lit).and_then(|d| {
+                    d.ok_or_else(|| {
+                        Error::new(
+                            ErrorKind::DataInvalid,
+                            "Transformed datum must not be 'None'",
```
I like this - thank you

marvinlanhenke commented on crates/iceberg/src/spec/transform.rs:
```diff
@@ -261,6 +270,174 @@ impl Transform {
+    ///
+    /// This allows predicates to be effectively applied to data
+    /// that has undergone transformation, enabling efficient querying
+    /// and filtering based on the original, untransformed data.
```
should I extend the docstring then? or doesn't the example specify this enough?

marvinlanhenke commented on crates/iceberg/src/spec/values.rs:
```diff
@@ -683,6 +684,47 @@ impl Datum {
     }
+
+    /// Create a new `Datum` with adjusted boundary for projection
+    pub fn boundary(&self, op: &PredicateOperator) -> Result<Option<Datum>> {
```
yes - but still implemented on `Datum` or as a helper fn on `Transform` with param: `&Datum` ? it feels strange to impl `Datum` across two modules.

marvinlanhenke commented:
@liurenjie1024 
I think I covered most of your suggestions. 
PTAL if the overall design and implementation in general is fine?

### Unresolved Issues:
- [ ] How to handle PrimitiveType::Time and PrimitiveType::Timestamp for `Dates` projection and boundary adjustment
- [ ] Move Transform specific logic into /transform (e.g. logic to handle Transform::Bucket projection -> /transform/bucket.rs)
- [ ] Port Java testsuite

to 1) I think we need to take PrimitiveType::Time and convert it to PrimitiveType::Date and then adjust the boundary. If this is correct, I still havent found a way to convert between those types. Perhaps someone can point me in the right direction?

---
*due to easter-holidays (and my two children) - I can probably do only some minimal work on this over the weekend

liurenjie1024 commented on crates/iceberg/src/spec/transform.rs:
```diff
@@ -384,6 +394,14 @@ impl Transform {
     }
 
+    /// Check if `Transform` is applicable on datum's `PrimitiveType`
+    fn can_transform(&self, datum: &Datum) -> Result<()> {
```
Why not just return bool directly?

liurenjie1024 commented on crates/iceberg/src/transform/mod.rs:
```diff
@@ -37,6 +36,15 @@ pub trait TransformFunction: Send {
     fn transform(&self, input: ArrayRef) -> Result<ArrayRef>;
     /// transform_literal will take an input literal and transform it into a new literal.
     fn transform_literal(&self, input: &Datum) -> Result<Option<Datum>>;
+    /// wrapper
```
```suggestion
    /// A thin wrapper around `transform_literal` to return error even when it's `None`.
```

liurenjie1024 commented:
Hi, @marvinlanhenke Thanks for your contribution, the overall design looks good to me! Please take your time and enjoy time with family😁

marvinlanhenke commented:
@Fokko @liurenjie1024 
I'm having trouble understanding/ verifiying the test-case for [timestamps_day_inclusive_epoch](https://github.com/apache/iceberg/blob/d350c9b8c995a2953aa8b80a0a1fc7cadc4dd16a/api/src/test/java/org/apache/iceberg/transforms/TestTimestampsProjection.java#L168) for the case `greater than`.

If I understood correctly timestamps are handled as [follows](https://github.com/apache/iceberg/blob/main/api/src/main/java/org/apache/iceberg/transforms/Timestamps.java#L134-L137): 
- call `truncateLong` -> add +1L, rewrite OP to GT_EQ and apply transformation
- then `fixInclusiveTimeProjection` -> for GT_EQ -> do nothing, simply return

so this would leave me with a predicate with an OP -> GT_EQ and a new literal (+1L) -> "1970-01-02T00:00:00.000000"?

Why would the test then check for "1970-01-01" // I can't find the part I'm missing; Is there some part that adjusts the boundary after the fact? Or is the +1L in truncateLong not applied? This is the last part I need to port.

Thanks for your help.

marvinlanhenke commented:
@Fokko @liurenjie1024 
PTAL

I ported all the tests - and fixed #311. Now, all tests are passing and align with the Java implementation.
I think we only can optimize the design/ structure by perhaps moving the test-suite to the respective transforms (e.g. bucket.rs, etc)? Other than that I think we are good to go?

liurenjie1024 commented:
> @Fokko @liurenjie1024 PTAL
> 
> I ported all the tests - and fixed #311. Now, all tests are passing and align with the Java implementation. I think we only can optimize the design/ structure by perhaps moving the test-suite to the respective transforms (e.g. bucket.rs, etc)? Other than that I think we are good to go?

Hi, @marvinlanhenke Thanks, I'll take a careful look of this pr. About moving the specific logic into respective transforms, do you plan to do it in this pr or in following prs?

marvinlanhenke commented:
> > @Fokko @liurenjie1024 PTAL
> > I ported all the tests - and fixed #311. Now, all tests are passing and align with the Java implementation. I think we only can optimize the design/ structure by perhaps moving the test-suite to the respective transforms (e.g. bucket.rs, etc)? Other than that I think we are good to go?
> 
> Hi, @marvinlanhenke Thanks, I'll take a careful look of this pr. About moving the specific logic into respective transforms, do you plan to do it in this pr or in following prs?

... I haven made up my mind yet. I think we can merge this if its okay, in order to unblock #253?

Also I'm not so sure anymore we have to move the logic at all? Since most of the helper functions handle logic that is not dependent of the type of transform? And don't know - perhaps you can outline the refactor high-level what you have in mind?

liurenjie1024 commented:
Hi, @marvinlanhenke I skimmed through the code and it seems that we only need to split tests into separate modules to make it easier to maintain and read. But I agree that we can merge it first to unblock following features, so it's up to you.

marvinlanhenke commented:
> Hi, @marvinlanhenke I skimmed through the code and it seems that we only need to split tests into separate modules to make it easier to maintain and read. But I agree that we can merge it first to unblock following features, so it's up to you.

Take your time while reviewing in the meantime - I'll open another draft with a possible refactor and we can compare whats best? I 'm thinking of splitting the code into Identity, Bucket, Truncate and Temporal logic - including the tests. This however might introduce some minor code duplication (I'll have to try and see for myself though)

liurenjie1024 commented on crates/iceberg/src/transform/temporal.rs:
```diff
@@ -269,11 +294,1995 @@ mod test {
+    }
+
+    #[test]
+    fn test_projection_timestamp_hour_lower_bound() -> Result<()> {
```
Really greatful for these test!

liurenjie1024 reviewed (commented):
Hi, @marvinlanhenke Thanks for pr, it looks great! I have some small suggestion to restructure the code to make it easier for review. Really greatful for these tests!

marvinlanhenke commented:
> Hi, @marvinlanhenke Thanks for pr, it looks great! I have some small suggestion to restructure the code to make it easier for review. Really greatful for these tests!

Thanks for the review, I'll get to your suggestions - those should be easy to fix.

marvinlanhenke commented on crates/iceberg/src/spec/transform.rs:
```diff
@@ -261,6 +269,300 @@ impl Transform {
+    }
+
+    /// Create a new `PredicateOperator`, rewritten for projection
+    fn projected_operator(
```
it applies to `Transform::Truncate` as well as the temporal transforms e.g. `Transform::Year` etc. I would leave the name as is since it applies to more than one transform?

liurenjie1024 replied:
Reasonable to me.

marvinlanhenke commented on crates/iceberg/src/spec/transform.rs:
```diff
@@ -261,6 +269,300 @@ impl Transform {
+    /// Returns Predicate with projection and possibly
+    /// rewritten `PredicateOperator`
+    fn transform_projected_boundary(
+        &self,
```
I need `&self` here since I call `fn adjust_projection(...)` which matches on `self` to decide wheter we `should_adjust` or not.

marvinlanhenke commented on crates/iceberg/src/spec/transform.rs:
```diff
@@ -261,6 +269,300 @@ impl Transform {
+        let func = create_transform_function(self)?;
+
+        let projection = match predicate {
+            BoundPredicate::Unary(expr) => match self {
```
I had the structure you suggested in an earlier version. I changed it the other way around since `predicate` has the smaller cardinality, which allows me to group more transforms into a single predicate match arm. I can change it back, however this would introduce more match arms and some code duplication?

liurenjie1024 replied:
I think some code duplication is worth so that we can have better readability?

marvinlanhenke replied:
sure, I'm already implementing it - since I wanted to compare for myself.

marvinlanhenke replied:
@liurenjie1024
I did a refactor changing the structure (matching order). I also extracted common functionality, renamed those helpers and updated the docs. I hope not only the structure but the overall design is more readable and understandable with those changes applied?

liurenjie1024 replied:
Yeah, it looks much better now, thanks! I'll take a careful review later.

liurenjie1024 commented on crates/iceberg/src/spec/transform.rs:
```diff
@@ -261,6 +269,323 @@ impl Transform {
+                new_set.insert(d);
+            };
+
+            new_set.insert(datum);
```
Should this be in else branch?

marvinlanhenke replied:
@liurenjie1024 
No. The original datum should always be inserted into the `new_set` alongside the adjusted_datum if we have any.
In Java it is done [here](https://github.com/apache/iceberg/blob/main/api/src/main/java/org/apache/iceberg/transforms/ProjectionUtil.java#L316-L319).

liurenjie1024 reviewed (approved):
Thanks @marvinlanhenke for this great pr with the  thorough  tests, really appreciate it! I have some small suggestions, but it looks great!

liurenjie1024 commented:
cc @Fokko Do you have other comments?

marvinlanhenke commented on crates/iceberg/src/spec/transform.rs:
```diff
@@ -261,6 +269,323 @@ impl Transform {
+
+    /// Projects a set expression to a predicate,
+    /// applying a transformation to each literal in the set.
+    fn project_set<T>(
```
fixed

marvinlanhenke commented on crates/iceberg/src/spec/transform.rs:
```diff
@@ -261,6 +269,323 @@ impl Transform {
+    /// This method evaluates a given binary expression and, if the operation
+    /// is equality (`Eq`) and the literal can be transformed, constructs a
+    /// `Predicate::Binary`variant representing the binary operation.
+    fn project_binary<T>(
```
fixed

marvinlanhenke commented:
> Hi, @marvinlanhenke Thanks for pr, it looks great! I have some small suggestion to restructure the code to make it easier for review. Really greatful for these tests!

Thanks for the review - I did some minor fixes according to your suggestions. If no other comments, I think we're good to go.

ZENOTME commented on crates/iceberg/src/spec/transform.rs:
```diff
@@ -261,6 +269,323 @@ impl Transform {
+
+    /// Adjust projection for temporal transforms, align with Java
+    /// implementation: https://github.com/apache/iceberg/blob/main/api/src/main/java/org/apache/iceberg/transforms/ProjectionUtil.java#L275
+    fn adjust_time_projection(
```
Seems this function is used to fix the bug from earlier iceberg. But I don't see same thing in pyiceberg transform. Do we really need this fix?  cc @liurenjie1024 @Fokko

ZENOTME replied:
> Is it breaking new things?

I have doubts because this does not exist in iceberg-python but it's a fixed logic. In java, also there is this process. I'm not sure whether we don't need it or iceberg-python miss it.
